### PR TITLE
Feature: educational videos

### PR DESCRIPTION
**Pull request by Basler182**
# *Feature: educational videos*

## :gear: Release Notes 
- education screen

image

image

image

## :white_check_mark: Testing
further tests will be added as soon as the basic architecture has been confirmed 

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by PSchmiedmayer — COMMENTED** [review]
Thank you @Basler182 for the improvements in this PR; great to see the color matching with the background for the menu bar & background! The white cards on the colored background look great!
Also great to see the lazy loading of the actual video views using the previews. I have two suggestions regarding the UI and would defer to @eldcn for a detailed code review.

1. I would suggest to use a 16:9 format for the placeholder views and use the full width for the previews. I would also suggest a small accent-color-based border around the view; made a small mockup based on your screenshots. Thanks for creating them! The play button should have the same size as now, just scaled with dragging the view frame. 
2. The detail screen looks a bit empty. Looking at this now I would rather suggest to replace the preview image with the video view and autostart it when the user has pressed the preview image and to stay on the same page. The lazy loading logic would still hold and this might nicely integrate with the current UI. Alternatively we should ask Alex if we have a longer description to show here that would justify a detail view.

344497426-b3f96580-4b32-40f1-a400-9f560a69a48e

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationScreen.kt
Comment (eldcn) [inline modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationScreen.kt:280]: If you recall a proposal of mine in the very beginning to move the previews and preview parameters in a separate dedicated package -> this was the reason, so that we could just exclude those file from code coverage. Maybe we can rethink this in the future (see this [issue](https://github.com/jacoco/jacoco/issues/1208) on jacocco)

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt
Comment (eldcn) [inline modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt:44]: Composables have generic names here, but those are bound to video related screen only (also the parameters take video). I would indicate this in the function names and mark them internal. Please also review nullability of the parameters after adapting `Video` model

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt
Comment (eldcn) [inline modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt:138]: I would get rid of the return value here and use:

```kotlin
@Composable
fun ExpandIcon(expanded: Boolean) {
    val vector = if (expanded) Icons.Filled.KeyboardArrowUp else Icons.Filled.KeyboardArrowDown
    Icon(
        imageVector = vector,
        contentDescription = null,
    )
}
```

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt
Comment (eldcn) [inline modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt:88]: Can't we just use `Colors.surface.copy(alpha = 0.9f)` instead?

File: modules/education/src/test/java/edu/stanford/spezi/modules/education/education/video/VideoViewModelTest.kt
Comment (eldcn) [inline modules/education/src/test/java/edu/stanford/spezi/modules/education/education/video/VideoViewModelTest.kt:38]: consider `every { navigator.navigateTo(EducationNavigationEvent.PopUp) } just Runs` in `setUp()` method

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/MainActivity.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/MainActivity.kt:157]: Great Idea!

File: core/design/src/main/kotlin/edu/stanford/spezi/core/design/theme/Elevations.kt
Comment (Basler182) [reply core/design/src/main/kotlin/edu/stanford/spezi/core/design/theme/Elevations.kt:25]: i thought it makes sense to have it standardised with small medium and large, so that if someone creates a new module they can be sure that they have access to all the other common sizes and also if someone external creates an app with the framework then they don't have to start defining their own sizes, but then i delete it for now

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/video/VideoScreen.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/video/VideoScreen.kt:50]: good idea . with a few small adjustments it was possible 🚀

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationViewModel.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationViewModel.kt:63]: yes, it would be great if we could automate this

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt:138]: Nice suggestion, thanks!

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt:88]: alpha 0.9 
image

vs
lighten colors
image

vs apha 0.2
image



i had tried that inital, but the problem is that the alpha value only changes the transparency of the colour, but does not actually make the colour brighter, so in my personal opinion it always looked less good. with a few colours the difference is not so big, but if you don't adjust the colours as well, you always have to set the alpha value much lower. with alpha 0.1, for example, you sometimes achieve similar optical results as with 0.9 if you also change the colours. 

for example:
alpha 0.7
image

vs lighten 0.7
image

Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt:88]: vs alpha 0.2 
image

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/education/EngageEducationRepository.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/education/EngageEducationRepository.kt:13]: Super idea, adjusted.

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/navigation/screens/AppScreen.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/navigation/screens/AppScreen.kt:56]: Great approach!

File: core/design/src/main/kotlin/edu/stanford/spezi/core/design/theme/Borders.kt
Comment (Basler182) [reply core/design/src/main/kotlin/edu/stanford/spezi/core/design/theme/Borders.kt:6]: that sounds good and we can do it like that. i have only included the elevations in Sizes now, because here i would assume, in contrast to possibly ratios, that this is needed again in many places

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/video/VideoViewModel.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/video/VideoViewModel.kt:21]: Yes, that makes a lot of sense

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationScreen.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationScreen.kt:120]: I had added another test while you were probably already reviewing :)

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableSection.kt:44]: good point. i originally intended it as a generic element too, but then more and more video logic was added and currently it doesn't seem like we want to reuse the composeable anywhere else, so i've now renamed it as you suggested and if it becomes relevant again later, i'll rebuild it to make it truly generic :)

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationViewModel.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationViewModel.kt:39]: I introduced a generic: 
```kotlin
sealed interface UiState<out T> {
    data object Loading : UiState<Nothing>
    data class Success<T>(val data: T) : UiState<T>
    data class Error(val message: String) : UiState<Nothing>
}```

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/video/VideoScreen.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/video/VideoScreen.kt:81]: Good Idea 🚀

File: core/design/src/main/kotlin/edu/stanford/spezi/core/design/component/TopAppBar.kt
Comment (Basler182) [reply core/design/src/main/kotlin/edu/stanford/spezi/core/design/component/TopAppBar.kt:13]: Good Idea, I'll add it

File: modules/education/src/androidTest/java/edu/stanford/spezi/modules/education/EducationModule.kt
Comment (Basler182) [reply modules/education/src/androidTest/java/edu/stanford/spezi/modules/education/EducationModule.kt:11]: Even while I use the parameterless version, Hilt complains that the module is missing. If I want to specify in the TestInstallIn what is to be replaced, I end up with the problem that the module is in the app, which is correct. However, we would actually prefer to leave the test in the module. How can we solve this elegantly?

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationScreen.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationScreen.kt:266]: if there is no other solution, we can do it like this 👍

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/education/EngageEducationRepository.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/education/EngageEducationRepository.kt:10]: the fact that only firebase specific things are used here is engage specific. in a final state, the account module would have to provide an interface, which can then be provided via a firebase module default, but alternatively e.g. AWS can also be used instead of firebase. at first i thought that since we use so many firebase specific, there will be few providers to replace it with, but [supabase](https://supabase.com/) would be a conceivable option, for example

File: core/utils/src/main/kotlin/edu/stanford/spezi/core/utils/extensions/SavedStateHandle.kt
Comment (eldcn) [inline core/utils/src/main/kotlin/edu/stanford/spezi/core/utils/extensions/SavedStateHandle.kt:6]: Great.

 After looking at the usage of the tests where we need to stub the json, I think extension function would cause some repetitive code. What do you think of wrapping this in a custom component:
 
 ```kotlin
 @ViewModelScoped
class RouteParamProvider @Inject constructor(
    val savedStateHandle: SavedStateHandle,
) {
    inline fun <reified T> decode(key: String): T {
        val jsonString = savedStateHandle.get<String>(key) ?: throw IllegalArgumentException("Argument not found")
        return Json.decodeFromString(jsonString)
    }
}
 ```
 
 This way we will just need to stub the return value of the decode method and no need to setup a json in our tests

File: modules/education/src/androidTest/java/edu/stanford/spezi/modules/education/EducationModule.kt
Comment (eldcn) [reply modules/education/src/androidTest/java/edu/stanford/spezi/modules/education/EducationModule.kt:11]: You can leave the replaces as empty

Also in your current `VideoScreenTest` you don't need  `@HiltAndroidTest`, `HiltAndroidRule`, and just use a `createComposeRule()`. Hilt setup is only needed in case you use hilt dependencies (view model in the parameterless composable). You can ignore for now and create a small issue for this to adjust

File: modules/education/src/androidTest/java/edu/stanford/spezi/modules/education/video/VideoScreenTest.kt
Comment (eldcn) [inline modules/education/src/androidTest/java/edu/stanford/spezi/modules/education/video/VideoScreenTest.kt:22]: Can't we just create a `Video` here which we pass in the screen below and avoid the factory and the custom state data class?

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationUiState.kt
Comment (eldcn) [inline modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationUiState.kt:20]: this can't be null anymore now, please also update the usages (reads via `description?.let`)

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableVideoSection.kt
Comment (eldcn) [inline modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/component/ExpandableVideoSection.kt:76]: Shouldn't be needed, `onExpand` updates the ui state which will cause a new composition. Or you can remove `expandedStartValue`, `onExpand` and `isExpanded` property from video section handle it in compose only with `var expanded by rememberSaveable { mutableStateOf(false) }`

**Review by eldcn — APPROVED** [review]
LGTM, well done 🚀  minor comments only. The main point open is testing, we can add those in a follow up to not block this PR

File: modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationUiState.kt
Comment (Basler182) [reply modules/education/src/main/java/edu/stanford/spezi/modules/education/videos/EducationUiState.kt:20]: Currently, the descriptions of the videos are still null. These will be added later. I have therefore temporarily left the null check in, but have already created a ticket in parallel so that we can think about revising these places later

File: modules/education/src/androidTest/java/edu/stanford/spezi/modules/education/video/VideoScreenTest.kt
Comment (Basler182) [reply modules/education/src/androidTest/java/edu/stanford/spezi/modules/education/video/VideoScreenTest.kt:22]: Yes this would also be possible. I noted it down and will exchange it later on

File: core/utils/src/main/kotlin/edu/stanford/spezi/core/utils/UiState.kt
Comment (Basler182) [reply core/utils/src/main/kotlin/edu/stanford/spezi/core/utils/UiState.kt:3]: It's a shame, I liked the approach because of the convention it offers with defined states. Even though it may not cover all cases, it could provide standardisation for a large number of them and introduce additional interfaces. i will remove it in an upcoming pr

**Pull request merged**
